shell/src/main/java/alluxio/cli/bundler/command/CollectLogCommand.java

```diff
@@ -53,15 +157,144 @@ public boolean hasSubCommand() {
+    mIncludedPrefix = new HashSet<>(FILE_NAMES);
+    // Define include list and exclude list
+    if (cl.hasOption(INCLUDE_OPTION_NAME)) {
+      Set<String> toInclude = parseFileNames(cl.getOptionValue(INCLUDE_OPTION_NAME));
+      System.out.format("Include the following filename prefixes: %s%n", toInclude);
+      mIncludedPrefix.addAll(toInclude);
+    }
+    if (cl.hasOption(EXCLUDE_OPTION_NAME)) {
+      mExcludedPrefix = parseFileNames(cl.getOptionValue(EXCLUDE_OPTION_NAME));
+      System.out.format("Exclude the following filename prefixes: %s%n", mExcludedPrefix);
+    }
+
+    // Check file timestamps
+    boolean checkTimeStamp = false;
+    if (cl.hasOption(START_OPTION_NAME)) {
+      String startTimeStr = cl.getOptionValue(START_OPTION_NAME);
+      mStartTime = parseDateTime(startTimeStr);
+      System.out.format("Time window start: %s%n", mStartTime);
+      checkTimeStamp = true;
+    }
+    if (cl.hasOption(END_OPTION_NAME)) {
+      String endTimeStr = cl.getOptionValue(END_OPTION_NAME);
+      mEndTime = parseDateTime(endTimeStr);
+      System.out.format("Time window end: %s%n", mEndTime);
+      checkTimeStamp = true;
+    }
+    if (mStartTime != null && mEndTime != null && mStartTime.isAfter(mEndTime)) {
+      System.err.format("ERROR: Start time %s is later than end time %s!%n",
+              mStartTime, mEndTime);
+    }
+
+    if (!mLogDir.exists()) {
+      System.err.format("ERROR: Alluxio log directory %s does not exist!%n", mLogDirPath);
+      return -1;
+    }
+
+    List<File> allFiles = CommonUtils.recursiveListLocalDir(mLogDir);
+    for (File f : allFiles) {
+      String relativePath = getRelativePathToLogDir(f);
+      try {
+        if (!shouldCopy(f, relativePath, checkTimeStamp)) {
+          continue;
+        }
+        File targetFile = new File(mWorkingDirPath, relativePath);
+        FileUtils.copyFile(f, targetFile, true);
+      } catch (IOException e) {
+        System.err.format("ERROR: file %s not found %s%n", f.getCanonicalPath(), e.getMessage());
+      }
+    }
 
     return 0;
   }
 
+  private String getRelativePathToLogDir(File f) {
+    return mLogDirUri.relativize(f.toURI()).getPath();
+  }
+
+  private boolean shouldCopy(File f, String relativePath, boolean checkTimeStamp)
+          throws IOException {
+    if (!fileNameIsWanted(relativePath)) {
+      return false;
+    }
+    if (checkTimeStamp) {
+      if (!fileTimeStampIsWanted(f)) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  private boolean fileNameIsWanted(String fileName) {
+    if (mExcludedPrefix != null) {
+      for (String x : mExcludedPrefix) {
+        if (fileName.startsWith(x)) {
+          return false;
+        }
+      }
+    }
+    for (String s : mIncludedPrefix) {
+      if (fileName.startsWith(s)) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private boolean fileTimeStampIsWanted(File f) throws IOException {
+    long timestamp = f.lastModified();
```

Comment:
We should guard this with `if (mEndTime != null)` and in the block, return false if it fails.

Reply:
I think since we don't require start time to have end time, we want it to default to true unless the file is filtered out with the specified time interval.

Note that the special case is we are not able to infer the start time of the file, we lean towards keeping it.

Updated the method description, hoping it makes the logic clearer.

Reply:
Ah gotcha. That makes sense to me. Updated